File: pilot/src/engine/events.rs
use chrono::Utc;
use serde::Serialize;

use crate::reference::{CheckpointFailure, ResponseFailure};

#[derive(Debug, Clone, Serialize)]
pub struct OpRecord {
    pub op_id: String,
    pub op_type: String,
    pub payload: serde_json::Value,
    pub status: String,
}

/// All event types written to events.jsonl.
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type")]
pub enum Event {
    RunStarted {
        run_id: String,
        hypothesis_id: String,
        timestamp: String,
    },
    AdapterDeployed {
        adapter: String,
        address: String,
        timestamp: String,
    },
    AdapterReady {
        adapter: String,
        timestamp: String,
    },
    BatchStarted {
        run_id: String,
        batch_id: String,
        offset: usize,
        size: usize,
        timestamp: String,
    },
    BatchCompleted {
        run_id: String,
        batch_id: String,
        duration_ms: u64,
        timestamp: String,
    },
    BatchFailed {
        run_id: String,
        batch_id: String,
        attempts: u32,
        error: String,
        timestamp: String,
    },
    ResponseFailed {
        run_id: String,
        batch_id: String,
        op_id: String,
        op: String,
        expected: String,
        actual: String,
        timestamp: String,
    },
    OperationBatch {
        run_id: String,
        batch_id: String,
        ops: Vec<OpRecord>,
        timestamp: String,
    },
    CheckpointStarted {
        run_id: String,
        checkpoint_id: String,
        keys: usize,
        timestamp: String,
    },
    CheckpointPassed {
        run_id: String,
        checkpoint_id: String,
        keys_checked: usize,
        timestamp: String,
    },
    CheckpointFailed {
        run_id: String,
        checkpoint_id: String,
        failures: usize,
        timestamp: String,
    },
    AdapterTeardown {
        adapter: String,
        timestamp: String,
    },
    RunCompleted {
        run_id: String,
        stop_reason: String,
        pass_rate: f64,
        timestamp: String,
    },
    RunStopped {
        run_id: String,
        stop_reason: String,
        timestamp: String,
    },
}

impl Event {
    pub fn now() -> String {
        Utc::now().format("%Y-%m-%dT%H:%M:%S%.3fZ").to_string()
    }

    pub fn operation_batch(run_id: &str, batch_id: &str, ops: Vec<OpRecord>) -> Self {
        Event::OperationBatch {
            run_id: run_id.to_string(),
            batch_id: batch_id.to_string(),
            ops,
            timestamp: Self::now(),
        }
    }

    pub fn run_started(run_id: &str, hypothesis_id: &str) -> Self {
        Event::RunStarted {
            run_id: run_id.to_string(),
            hypothesis_id: hypothesis_id.to_string(),
            timestamp: Self::now(),
        }
    }

    pub fn batch_started(run_id: &str, batch_id: &str, offset: usize, size: usize) -> Self {
        Event::BatchStarted {
            run_id: run_id.to_string(),
            batch_id: batch_id.to_string(),
            offset,
            size,
            timestamp: Self::now(),
        }
    }

    pub fn batch_completed(run_id: &str, batch_id: &str, duration_ms: u64) -> Self {
        Event::BatchCompleted {
            run_id: run_id.to_string(),
            batch_id: batch_id.to_string(),
            duration_ms,
            timestamp: Self::now(),
        }
    }

    pub fn batch_failed(run_id: &str, batch_id: &str, attempts: u32, error: &str) -> Self {
        Event::BatchFailed {
            run_id: run_id.to_string(),
            batch_id: batch_id.to_string(),
            attempts,
            error: error.to_string(),
            timestamp: Self::now(),
        }
    }

    pub fn response_failed(run_id: &str, batch_id: &str, failure: &ResponseFailure) -> Self {
        Event::ResponseFailed {
            run_id: run_id.to_string(),
            batch_id: batch_id.to_string(),
            op_id: failure.op_id.clone(),
            op: failure.op.clone(),
            expected: failure.expected.clone(),
            actual: failure.actual.clone(),
            timestamp: Self::now(),
        }
    }

    pub fn checkpoint_started(run_id: &str, checkpoint_id: &str, keys: usize) -> Self {
        Event::CheckpointStarted {
            run_id: run_id.to_string(),
            checkpoint_id: checkpoint_id.to_string(),
            keys,
            timestamp: Self::now(),
        }
    }

    pub fn checkpoint_passed(run_id: &str, checkpoint_id: &str, keys_checked: usize) -> Self {
        Event::CheckpointPassed {
            run_id: run_id.to_string(),
            checkpoint_id: checkpoint_id.to_string(),
            keys_checked,
            timestamp: Self::now(),
        }
    }

    pub fn checkpoint_failed(run_id: &str, checkpoint_id: &str, failures: usize) -> Self {
        Event::CheckpointFailed {
            run_id: run_id.to_string(),
            checkpoint_id: checkpoint_id.to_string(),
            failures,
            timestamp: Self::now(),
        }
    }

    pub fn run_completed(run_id: &str, stop_reason: &str, pass_rate: f64) -> Self {
        Event::RunCompleted {
            run_id: run_id.to_string(),
            stop_reason: stop_reason.to_string(),
            pass_rate,
            timestamp: Self::now(),
        }
    }

    pub fn run_stopped(run_id: &str, stop_reason: &str) -> Self {
        Event::RunStopped {
            run_id: run_id.to_string(),
            stop_reason: stop_reason.to_string(),
            timestamp: Self::now(),
        }
    }

    pub fn to_json(&self) -> String {
        serde_json::to_string(self).unwrap_or_default()
    }
}

File: pilot/engine/events.go
package engine

import (
	"encoding/json"
	"fmt"
	"time"
)

// OpRecord captures the result of a single operation within a batch.
type OpRecord struct {
	OpID     string      `json:"op_id"`
	OpType   string      `json:"op_type"`
	Payload  interface{} `json:"payload"`
	Status   string      `json:"status"`
	Response interface{} `json:"response"`
	Expected interface{} `json:"expected,omitempty"`
	Actual   interface{} `json:"actual,omitempty"`
	Verified *bool       `json:"verified,omitempty"`
}

// CheckpointDetail records the comparison result for a single key.
type CheckpointDetail struct {
	Key      string      `json:"key"`
	Matched  bool        `json:"matched"`
	Expected interface{} `json:"expected"`
	Actual   interface{} `json:"actual"`
}

// EventType enumerates the kinds of events.
type EventType string

const (
	EventTypeRunStarted      EventType = "RunStarted"
	EventTypeOperationBatch  EventType = "OperationBatch"
	EventTypeBatchFailed     EventType = "BatchFailed"
	EventTypeSafetyViolation EventType = "SafetyViolation"
	EventTypeCheckpoint      EventType = "Checkpoint"
	EventTypeRunCompleted    EventType = "RunCompleted"
	EventTypeRunStopped      EventType = "RunStopped"
)

// Event represents a single event in the events.jsonl stream.
type Event struct {
	Type         EventType          `json:"type"`
	RunID        string             `json:"run_id"`
	Timestamp    string             `json:"timestamp"`
	HypothesisID string             `json:"hypothesis_id,omitempty"`
	BatchID      string             `json:"batch_id,omitempty"`
	Offset       int                `json:"offset,omitempty"`
	Size         int                `json:"size,omitempty"`
	DurationMs   uint64             `json:"duration_ms,omitempty"`
	Ops          []OpRecord         `json:"ops,omitempty"`
	Attempts     uint32             `json:"attempts,omitempty"`
	Error        string             `json:"error,omitempty"`
	OpID         string             `json:"op_id,omitempty"`
	Op           string             `json:"op,omitempty"`
	Expected     string             `json:"expected,omitempty"`
	Actual       string             `json:"actual,omitempty"`
	CheckpointID string             `json:"checkpoint_id,omitempty"`
	Keys         int                `json:"keys,omitempty"`
	Passed       *bool              `json:"passed,omitempty"`
	Details      []CheckpointDetail `json:"details,omitempty"`
	StopReason   string             `json:"stop_reason,omitempty"`
	PassRate     *float64           `json:"pass_rate,omitempty"`
}

// Now returns the current UTC time in ISO8601 format with milliseconds.
func Now() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05.000Z")
}

// NewRunStartedEvent creates a RunStarted event.
func NewRunStartedEvent(runID, hypothesisID string) Event {
	return Event{
		Type:         EventTypeRunStarted,
		RunID:        runID,
		HypothesisID: hypothesisID,
		Timestamp:    Now(),
	}
}

// NewOperationBatchEvent creates an OperationBatch event.
func NewOperationBatchEvent(runID, batchID string, offset int, durationMs uint64, ops []OpRecord) Event {
	return Event{
		Type:       EventTypeOperationBatch,
		RunID:      runID,
		BatchID:    batchID,
		Offset:     offset,
		Size:       len(ops),
		DurationMs: durationMs,
		Ops:        ops,
		Timestamp:  Now(),
	}
}

// NewBatchFailedEvent creates a BatchFailed event.
func NewBatchFailedEvent(runID, batchID string, attempts uint32, errMsg string) Event {
	return Event{
		Type:      EventTypeBatchFailed,
		RunID:     runID,
		BatchID:   batchID,
		Attempts:  attempts,
		Error:     errMsg,
		Timestamp: Now(),
	}
}

// NewSafetyViolationEvent creates a SafetyViolation event.
func NewSafetyViolationEvent(runID, batchID, opID, op, expected, actual string) Event {
	return Event{
		Type:      EventTypeSafetyViolation,
		RunID:     runID,
		BatchID:   batchID,
		OpID:      opID,
		Op:        op,
		Expected:  expected,
		Actual:    actual,
		Timestamp: Now(),
	}
}

// NewCheckpointEvent creates a Checkpoint event.
func NewCheckpointEvent(runID, checkpointID string, durationMs uint64, passed bool, details []CheckpointDetail) Event {
	keys := len(details)
	return Event{
		Type:         EventTypeCheckpoint,
		RunID:        runID,
		CheckpointID: checkpointID,
		Keys:         keys,
		Passed:       &passed,
		DurationMs:   durationMs,
		Details:      details,
		Timestamp:    Now(),
	}
}

// NewRunCompletedEvent creates a RunCompleted event.
func NewRunCompletedEvent(runID, stopReason string, passRate float64) Event {
	return Event{
		Type:       EventTypeRunCompleted,
		RunID:      runID,
		StopReason: stopReason,
		PassRate:   &passRate,
		Timestamp:  Now(),
	}
}

// NewRunStoppedEvent creates a RunStopped event.
func NewRunStoppedEvent(runID, stopReason string) Event {
	return Event{
		Type:       EventTypeRunStopped,
		RunID:      runID,
		StopReason: stopReason,
		Timestamp:  Now(),
	}
}

// ToJSON serializes the event to a JSON string.
func (e *Event) ToJSON() string {
	data, err := json.Marshal(e)
	if err != nil {
		return fmt.Sprintf(`{"type":"error","message":"%s"}`, err.Error())
	}
	return string(data)
}
